core/blockchain.go

```diff
@@ -258,12 +258,27 @@ type BlockChain struct {
 	forker     *ForkChoice
 	vmConfig   vm.Config
 	logger     *tracing.Hooks
+
+	// true if witness trie nodes will be prefetched during block execution. The prefetcher will also terminate
+	// synchronously in this case.
+	collectWitness bool
 }
 
 // NewBlockChain returns a fully initialised block chain using information
 // available in the database. It initialises the default Ethereum Validator
 // and Processor.
 func NewBlockChain(db ethdb.Database, cacheConfig *CacheConfig, genesis *Genesis, overrides *ChainOverrides, engine consensus.Engine, vmConfig vm.Config, shouldPreserve func(header *types.Header) bool, txLookupLimit *uint64) (*BlockChain, error) {
+	return newBlockChain(db, cacheConfig, genesis, overrides, engine, vmConfig, shouldPreserve, txLookupLimit, false)
+}
+
+// NewBlockChainWithWitnessCollection returns a fully initialized BlockChain, Validator and Processor the same as
+// NewBlockchain.  However, the returned Blockchain instance is configured to collect all necessary account/storage trie
+// nodes necessary to build a stateless witness for any executed block.  It is a base for future functionality, and does
+// not do anything on its own.
+func NewBlockChainWithWitnessCollection(db ethdb.Database, cacheConfig *CacheConfig, genesis *Genesis, overrides *ChainOverrides, engine consensus.Engine, vmConfig vm.Config, shouldPreserve func(header *types.Header) bool, txLookupLimit *uint64) (*BlockChain, error) {
+	return newBlockChain(db, cacheConfig, genesis, overrides, engine, vmConfig, shouldPreserve, txLookupLimit, true)
+}
+func newBlockChain(db ethdb.Database, cacheConfig *CacheConfig, genesis *Genesis, overrides *ChainOverrides, engine consensus.Engine, vmConfig vm.Config, shouldPreserve func(header *types.Header) bool, txLookupLimit *uint64, collectWitnesses bool) (*BlockChain, error) {
```

Comment:
This is IMO a very bad API to duplicate the constructor just to have that one flag in there. Can't we make it part of one of the configs? E.g. the vm config?